server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,438 @@
+                        decommissionAttribute.toString()
+                    ),
+                    e
+                );
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                DecommissionAttributeMetadata decommissionAttributeMetadata = newState.metadata()
+                    .custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttribute.equals(decommissionAttributeMetadata.decommissionAttribute());
+                decommissionClusterManagerNodes(decommissionAttributeMetadata.decommissionAttribute(), listener);
+            }
+        });
+    }
+
+    private synchronized void decommissionClusterManagerNodes(
+        final DecommissionAttribute decommissionAttribute,
+        ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+        Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned = filterNodesWithDecommissionAttribute(state, decommissionAttribute, true);
+        // This check doesn't seem to be needed as exclusion automatically shrinks the config before sending the response.
+        // We can guarantee that because of exclusion there wouldn't be a quorum loss and if the service gets a successful response,
+        // we are certain that the config is updated and nodes are ready to be kicked out.
+        // Please add comment if you feel there could be a edge case here.
+        // try {
+        // // this is a sanity check that the cluster will not go into a quorum loss state because of exclusion
+        // ensureNoQuorumLossDueToDecommissioning(
+        // decommissionAttribute,
+        // clusterManagerNodesToBeDecommissioned,
+        // state.getLastCommittedConfiguration()
+        // );
+        // } catch (DecommissioningFailedException dfe) {
+        // listener.onFailure(dfe);
+        // decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.FAILED, statusUpdateListener());
+        // return;
+        // }
+
+        ActionListener<Void> exclusionListener = new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void unused) {
+                if (transportService.getLocalNode().isClusterManagerNode()
+                    && !nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
+                    logger.info("will attempt to fail decommissioned nodes as local node is eligible to process the request");
+                    // we are good here to send the response now as the request is processed by an eligible active leader
+                    // and to-be-decommissioned cluster manager is no more part of Voting Configuration
+                    listener.onResponse(new ClusterStateUpdateResponse(true));
```

Comment:
the user request would have ended here or when onFailure() is triggered, is that right?

Reply:
Yes, whenever the exclusion is processed and the further request is executed by eligible leader, we are sending the response back to the user and further process is run in the background